File: backend/services/lesson_service.py
from extensions import db
from repositories.lesson_repository import LessonRepository
from errors import NotFoundError

lesson_repo = LessonRepository()


def get_lessons(teacher_id: int, q: str = "", page: int = 1, limit: int = 10) -> dict:
    result = lesson_repo.find_by_teacher(teacher_id, q, page, limit)
    for r in result["items"]:
        if "created_at" in r and r["created_at"]:
            r["created_at"] = str(r["created_at"])
    return result


def get_lesson(lesson_id: int) -> dict:
    lesson = lesson_repo.find_by_id(lesson_id)
    if not lesson:
        raise NotFoundError("Bài học không tồn tại")
    exams = [
        {
            "id": e.id, "name": e.name, "description": e.description,
            "date_created": str(e.date_created) if e.date_created else None,
        }
        for e in lesson.exams
    ]
    return {
        "lessonId": lesson.id, "lessonTitle": lesson.title,
        "description": lesson.description or "", "exams": exams,
    }


def create_lesson(teacher_id: int, title: str) -> None:
    lesson_repo.create(teacher_id, title)
    db.session.commit()


def update_lesson(lesson_id: int, teacher_id: int, title: str) -> None:
    lesson = lesson_repo.find_by_id(lesson_id)
    if not lesson or lesson.teacher_id != teacher_id:
        raise NotFoundError("Bài học không tồn tại")
    lesson_repo.update(lesson, title)
    db.session.commit()


def delete_lesson(lesson_id: int, teacher_id: int) -> None:
    lesson = lesson_repo.find_by_id(lesson_id)
    if not lesson or lesson.teacher_id != teacher_id:
        raise NotFoundError("Bài học không tồn tại")
    lesson_repo.delete(lesson)
    db.session.commit()

File: backend/src/services/lesson.service.js
import * as lessonRepo from '../repositories/lesson.repository.js';
import { NotFoundError } from '../utils/error.utils.js';
import { formatDate } from '../utils/date.utils.js';

export const getLessons = async (teacherId, q = '', page = 1, limit = 10) => {
  const result = await lessonRepo.findByTeacher(teacherId, q, page, limit);
  result.items = result.items.map(r => ({
    ...r,
    created_at: formatDate(r.created_at),
  }));
  return result;
};

export const getLesson = async (lessonId) => {
  const lesson = await lessonRepo.findById(lessonId);
  if (!lesson) throw new NotFoundError('Bài học không tồn tại');
  const exams = await lessonRepo.findExamsByLesson(lessonId);
  return {
    lessonId: lesson.id,
    lessonTitle: lesson.title,
    description: lesson.description || '',
    exams: exams.map(e => ({
      id: e.id, name: e.name, description: e.description,
      date_created: formatDate(e.date_created),
    })),
  };
};

export const createLesson = async (teacherId, title) => {
  await lessonRepo.createLesson(teacherId, title);
};

export const updateLesson = async (lessonId, teacherId, title) => {
  const lesson = await lessonRepo.findById(lessonId);
  if (!lesson || lesson.teacher_id !== teacherId) throw new NotFoundError('Bài học không tồn tại');
  await lessonRepo.updateLesson(lessonId, title);
};

export const deleteLesson = async (lessonId, teacherId) => {
  const lesson = await lessonRepo.findById(lessonId);
  if (!lesson || lesson.teacher_id !== teacherId) throw new NotFoundError('Bài học không tồn tại');
  await lessonRepo.deleteLesson(lessonId);
};
